pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java

```diff
@@ -489,12 +489,16 @@ public boolean index(GenericRow row, @Nullable RowMetadata rowMetadata)
     if (isUpsertEnabled()) {
       RecordInfo recordInfo = getRecordInfo(row, numDocsIndexed);
       GenericRow updatedRow = _partitionUpsertMetadataManager.updateRecord(row, recordInfo);
-      updateDictionary(updatedRow);
-      addNewRow(numDocsIndexed, updatedRow);
-      // Update number of documents indexed before handling the upsert metadata so that the record becomes queryable
-      // once validated
-      canTakeMore = numDocsIndexed++ < _capacity;
-      _partitionUpsertMetadataManager.addRecord(this, recordInfo);
+      // if record doesn't need to be dropped, then persist in segment and update metadata hashmap
+      if (!_partitionUpsertMetadataManager.shouldDropRecord(recordInfo)) {
```

Comment:
This check is already performed in `doAddRecord()`. Suggest modifying `addRecord()` to return `boolean` which indicates whether the record should be kept. Then we may modify this part into
```suggestion
      if (_partitionUpsertMetadataManager.addRecord(this, recordInfo)) {
```

Reply:
Hey @Jackie-Jiang at present `_partitionUpsertMetadataManager.addRecord` is called after `addNewRow`. Do you foresee any issues in changing the order here?

Reply:
One scenario I can think of where this might cause an issue is that we add a record in the metadata hashmap but somehow `addNewRow` fails due to some indexing issues. At present, it throws an exception which doesn't call the `_partitionUpsertMetadataManager.addRecord` as well which makes sense.

Reply:
I checked and there should be no issue changing the order because the row is not added to the metadata manager if it is out of order

Reply:
> I checked and there should be no issue changing the order because the row is not added to the metadata manager if it is out of order

What about a row which is not out-of-order? It gets added to metadata manager but `addNewRow` call fails due to some issue. Will that not cause issues when further updates are received for that record?

Reply:
If it throws exception, in both cases `_numDocsIndexed` won't be updated, and we will end up re-indexing the same docId. That needs to be solved separately, but very good point!

Reply:
Umm i think there can be a race-condition something like:

- `Key: A, Current Segment: S1, Valid Doc ID: 2`  and say `addNewRow` call fails for this.
MetadataMap: A -> RecordLocation (S1, 2)
- `Key: B, Current Segment: S1, Valid Doc ID: 2` (This also gets valid Doc ID as 2 as `_numDocsIndexed` didn't update). 
MetadataMap:  A -> RecordLocation (S1, 2), B -> RecordLocation (S1, 2)
- Now say, we receive a record for key A again after sometime when we are processing segment S2.
In that scenario, we will invoke a removeDocID command for (S1, 2) here - https://github.com/apache/pinot/blob/51335a2db1720ac18da8818a4b565b557a0c667b/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java#L264
which will invalidate docID for B and will stop responding it in queries. 

Did I miss anything? 😅

Reply:
The analysis is correct. What I meant is that currently it is anyway not handled correctly, and we should solve this problem in a separate PR. Assuming `addNewRow()` failed, dictionary was already updated, and the first several columns might also be added already. Currently it works because that call never fails. The proper fix should be to ensure even if it throws exception, we should continue on other columns (e.g. put a default value for the failed column), and increase the doc id anyway.

Reply:
Sounds good to me! Updated the patch based on this.